File: java-workshop/part-2-light-mocks/src/main/java/service/PricingService.java
package service;

import domain.DiscountCode;
import domain.OrderItem;

import java.util.List;

public class PricingService {
    public static final double TAX_RATE = 0.20;
    public static final double FREE_SHIPPING_THRESHOLD = 100.0;
    public static final double SHIPPING_COST = 5.00;

    public double calculateTotal(List<OrderItem> items) {
        double subtotal = calculateSubtotal(items);
        double tax = calculateTax(subtotal);
        double shipping = calculateShipping(subtotal);
        return subtotal + tax + shipping;
    }

    public double calculateTotal(List<OrderItem> items, DiscountCode discountCode) {
        double subtotal = calculateSubtotal(items);
        double discountedSubtotal = applyDiscount(subtotal, discountCode);
        double tax = calculateTax(discountedSubtotal);
        double shipping = calculateShipping(discountedSubtotal);
        return discountedSubtotal + tax + shipping;
    }

    private double calculateSubtotal(List<OrderItem> items) {
        return items.stream()
                .mapToDouble(item -> item.quantity() * item.unitPrice())
                .sum();
    }

    private double applyDiscount(double subtotal, DiscountCode discountCode) {
        double discountFactor = 1.0 - (discountCode.getPercentage() / 100.0);
        return subtotal * discountFactor;
    }

    private double calculateTax(double subtotal) {
        return subtotal * TAX_RATE;
    }

    private double calculateShipping(double subtotal) {
        return subtotal >= FREE_SHIPPING_THRESHOLD ? 0.0 : SHIPPING_COST;
    }
}

File: csharp-workshop/Part1.StaticDeps/Service/PricingService.cs
using Part1.StaticDeps.Domain;

namespace Part1.StaticDeps.Service;

/// <summary>
/// Static utility class for price calculations.
/// </summary>
public static class PricingService
{
    public const double TaxRate = 0.20;
    public const double FreeShippingThreshold = 100.0;
    public const double ShippingCost = 5.00;

    public static double CalculateTotal(IReadOnlyList<OrderItem> items)
    {
        double subtotal = CalculateSubtotal(items);
        double tax = CalculateTax(subtotal);
        double shipping = CalculateShipping(subtotal);
        return subtotal + tax + shipping;
    }

    public static double CalculateTotal(IReadOnlyList<OrderItem> items, DiscountCode discountCode)
    {
        double subtotal = CalculateSubtotal(items);
        double discountedSubtotal = ApplyDiscount(subtotal, discountCode);
        double tax = CalculateTax(discountedSubtotal);
        double shipping = CalculateShipping(discountedSubtotal);
        return discountedSubtotal + tax + shipping;
    }

    private static double CalculateSubtotal(IReadOnlyList<OrderItem> items) =>
        items.Sum(item => item.Quantity * item.UnitPrice);

    private static double ApplyDiscount(double subtotal, DiscountCode discountCode)
    {
        double discountFactor = 1.0 - (discountCode.GetPercentage() / 100.0);
        return subtotal * discountFactor;
    }

    private static double CalculateTax(double subtotal) =>
        subtotal * TaxRate;

    private static double CalculateShipping(double subtotal) =>
        subtotal >= FreeShippingThreshold ? 0.0 : ShippingCost;
}
